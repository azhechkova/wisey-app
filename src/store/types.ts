import { TrackVideoType } from '~/types';

export type RootState = {
  app: {
    activeLesson: string;
    videoProgress: TrackVideoType[];
  };
};
