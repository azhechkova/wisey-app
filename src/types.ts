type MetaObject = {
  slug?: string;
  skills?: string[];
  courseVideoPreview?: {
    link?: string;
    previewImageLink?: string;
    duration?: number;
  };
};
export type LessonType = {
  id: string;
  title: string;
  duration: number;
  order: number;
  type: 'video';
  status: 'unlocked' | 'locked';
  link: string;
  previewImageLink: string;
  meta?: MetaObject | null;
};

export type CourseType = {
  id: string;
  title: string;
  tags: string[];
  launchDate?: string;
  status: 'launched';
  description: string;
  duration: number;
  lessonsCount?: number;
  lessons: LessonType[];
  containsLockedLessons: boolean;
  previewImageLink: string;
  rating: number;
  meta?: MetaObject | null;
};

export type ErrorType = Error | unknown | null;

export type TApiResponse<T> = {
  loading: boolean;
  error: ErrorType;
  data: T | null;
};

export type TrackVideoType = {
  videoId: string;
  progress: number;
  isFinished: boolean;
};
