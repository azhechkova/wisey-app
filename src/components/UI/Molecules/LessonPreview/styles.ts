import { makeStyles } from 'tss-react/mui';

interface StyleProps {
  isLocked: boolean;
  isActive?: boolean;
}

const useStyles = makeStyles<StyleProps>()((theme, { isLocked, isActive }) => ({
  root: {
    padding: '0.75rem 1rem',
    borderRadius: '0.75rem',
    display: 'flex',
    width: '100%',
    gap: '0.75rem',
    border: '2px solid',
    borderColor: isActive ? theme.palette.secondary.main : 'transparent',
    backgroundColor: '#fff',
    cursor: 'pointer',
    opacity: isLocked ? '0.5' : 1,
  },
  info: {
    textAlign: 'left',
  },
  icon: {
    height: '48px',
    width: '48px',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  title: {
    fontSize: '0.75rem',
    fontWeight: 500,
  },
}));

export default useStyles;
