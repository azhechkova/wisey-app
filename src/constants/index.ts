export const PROJECT_NAME = 'wisey-app';

export const VIDEOS_STORAGE = 'wisey-app-videos';

export const LESSON_STATUS = {
  locked: 'locked',
  unlocked: 'unlocked',
};

const HOST = process.env.REACT_APP_API_HOST;
const VERSION = process.env.REACT_APP_API_VERSION;

export const baseURL = `${HOST}/${VERSION}`;
