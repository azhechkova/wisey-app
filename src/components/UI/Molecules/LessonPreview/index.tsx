import React from 'react';
import { Box, Typography } from '@mui/material';
import { Slideshow, LockOutlined, Check } from '@mui/icons-material';

import { LessonType } from '~/types';
import { LESSON_STATUS } from '~/constants';

import Duration from '~/components/UI/Atoms/Duration';

import useStyles from './styles';

interface LessonPreviewProps {
  lesson: LessonType;
  onLessonChange: (lessonId: string) => void;
  state: {
    isActive?: boolean;
    isFinished?: boolean;
  };
}

const LessonPreview = ({
  lesson,
  onLessonChange,
  state: { isFinished, isActive },
}: LessonPreviewProps): JSX.Element => {
  const styleProps = {
    isLocked: lesson?.status === LESSON_STATUS.locked,
    isActive,
    isFinished,
  };
  const { classes } = useStyles(styleProps);

  const onClick = (): void => {
    onLessonChange(lesson.id);
  };

  return (
    <button
      aria-label="lesson button"
      onClick={onClick}
      type="button"
      className={classes.root}
    >
      <Box className={classes.icon}>
        {lesson.status === LESSON_STATUS.unlocked ? (
          <Slideshow />
        ) : (
          <LockOutlined />
        )}
      </Box>
      <Box className={classes.info}>
        <Typography
          className={classes.title}
          component="h4"
          variant="subtitle1"
        >
          {lesson.title}
        </Typography>
        <Duration duration={lesson.duration} />
      </Box>
      {isFinished && <Check className={classes.checkIcon} />}
    </button>
  );
};

export default LessonPreview;
