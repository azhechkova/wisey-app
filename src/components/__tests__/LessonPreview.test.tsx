import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { LessonType } from '~/types';
import { LESSON_STATUS } from '~/constants';
import LessonPreview from '../UI/Molecules/LessonPreview';

describe('LessonPreview component', () => {
  const lesson = {
    id: '1',
    title: 'Test Lesson',
    duration: 60,
    status: LESSON_STATUS.unlocked,
  } as LessonType;
  const onLessonChange = jest.fn();

  it('should render the lesson title and duration', () => {
    const { getByText } = render(
      <LessonPreview
        lesson={lesson}
        onLessonChange={onLessonChange}
        state={{}}
      />,
    );

    expect(getByText('Test Lesson')).toBeInTheDocument();
    expect(getByText('01:00 min')).toBeInTheDocument();
  });

  it('should call onLessonChange when clicked', () => {
    const { getByLabelText } = render(
      <LessonPreview
        lesson={lesson}
        onLessonChange={onLessonChange}
        state={{}}
      />,
    );

    fireEvent.click(getByLabelText('lesson button'));
    expect(onLessonChange).toHaveBeenCalledWith('1');
  });

  it('should render the Slideshow icon when the lesson is unlocked', () => {
    const { container } = render(
      <LessonPreview
        lesson={lesson}
        onLessonChange={onLessonChange}
        state={{}}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render the LockOutlined icon when the lesson is locked', () => {
    const lockedLesson = {
      id: '2',
      title: 'Locked Lesson',
      duration: 30,
      status: LESSON_STATUS.locked,
    } as LessonType;
    const { container } = render(
      <LessonPreview
        lesson={lockedLesson}
        onLessonChange={onLessonChange}
        state={{}}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render the Check icon when the lesson is finished', () => {
    const { container } = render(
      <LessonPreview
        lesson={lesson}
        onLessonChange={onLessonChange}
        state={{ isFinished: true }}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
