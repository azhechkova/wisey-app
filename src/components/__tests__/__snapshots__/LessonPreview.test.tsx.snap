// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LessonPreview component should render the Check icon when the lesson is finished 1`] = `
<div>
  <button
    aria-label="lesson button"
    class="css-zjoe5b-root"
    type="button"
  >
    <div
      class="MuiBox-root css-1rl4y9l-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SlideshowIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 8v8l5-4-5-4zm9-5H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root css-1ur0ipg-info"
    >
      <h4
        class="MuiTypography-root MuiTypography-subtitle1 css-1sdlpag-MuiTypography-root-title"
      >
        Test Lesson
      </h4>
      <div
        class="MuiBox-root css-vot3lh-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-5w6cp7-MuiSvgIcon-root-icon"
          data-testid="WatchLaterIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.5 2 2 6.5 2 12s4.5 10 10 10 10-4.5 10-10S17.5 2 12 2zm4.2 14.2L11 13V7h1.5v5.2l4.5 2.7-.8 1.3z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1djztph-MuiTypography-root-text"
        >
          01:00 min
        </p>
      </div>
    </div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-8253z7-MuiSvgIcon-root-checkIcon"
      data-testid="CheckIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
      />
    </svg>
  </button>
</div>
`;

exports[`LessonPreview component should render the LockOutlined icon when the lesson is locked 1`] = `
<div>
  <button
    aria-label="lesson button"
    class="css-bsistv-root"
    type="button"
  >
    <div
      class="MuiBox-root css-1rl4y9l-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="LockOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root css-1ur0ipg-info"
    >
      <h4
        class="MuiTypography-root MuiTypography-subtitle1 css-1sdlpag-MuiTypography-root-title"
      >
        Locked Lesson
      </h4>
      <div
        class="MuiBox-root css-vot3lh-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-5w6cp7-MuiSvgIcon-root-icon"
          data-testid="WatchLaterIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.5 2 2 6.5 2 12s4.5 10 10 10 10-4.5 10-10S17.5 2 12 2zm4.2 14.2L11 13V7h1.5v5.2l4.5 2.7-.8 1.3z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1djztph-MuiTypography-root-text"
        >
          00:30 min
        </p>
      </div>
    </div>
  </button>
</div>
`;

exports[`LessonPreview component should render the Slideshow icon when the lesson is unlocked 1`] = `
<div>
  <button
    aria-label="lesson button"
    class="css-2vvkyg-root"
    type="button"
  >
    <div
      class="MuiBox-root css-1rl4y9l-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SlideshowIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 8v8l5-4-5-4zm9-5H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root css-1ur0ipg-info"
    >
      <h4
        class="MuiTypography-root MuiTypography-subtitle1 css-1sdlpag-MuiTypography-root-title"
      >
        Test Lesson
      </h4>
      <div
        class="MuiBox-root css-vot3lh-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-5w6cp7-MuiSvgIcon-root-icon"
          data-testid="WatchLaterIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.5 2 2 6.5 2 12s4.5 10 10 10 10-4.5 10-10S17.5 2 12 2zm4.2 14.2L11 13V7h1.5v5.2l4.5 2.7-.8 1.3z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1djztph-MuiTypography-root-text"
        >
          01:00 min
        </p>
      </div>
    </div>
  </button>
</div>
`;
